### DISCUSSION TITLE
Tailwind Fails to Apply Dynamic Hover Color Effect

### DISCUSSION THREAD

**Discussion by emiliaiwu** [tailwindlabs/tailwindcss#12138] [Help]
I'm experiencing an issue with Tailwind CSS where hover effects for a background color (bg-[color]) are not rendering correctly when I pass the color dynamically. However, when I hardcode the color directly as bg-[#fffff], it works as expected.

I have a React component where I want to apply a hover effect based on a dynamically generated color value. Here's my code snippet:

```
const color = userPreferences.color;
const hoverColorClass = `bg-[${color}]`;

return (
  <li
    key={index}
    className={`hover:${hoverColorClass} h-10 font-DMSans flex items-center justify-center relative group`}
  >
    {/* My list item content */}
  </li>
);
```

In this code:

userPreferences.color contains a color value with quotes (e.g., "#ffffff").
I construct the hoverColorClass string with square brackets around ${color} to match the Tailwind CSS class format bg-[color].
I apply the hover class using the className prop with hover:${hoverColorClass}.
However, the hover color effect is not being applied as expected when I hover over the list item.

**Comment by alex-krasikau** [comment]
Hello @emiliaiwu, thanks for your question!

The issue here is that you are using a template string to construct the resulting class. Tailwind works by parsing your raw code files and finding class appearances. So if it doesn't find `hover:bg-[#ffffff]` anywhere in your code base, it will not generate the corresponding class declaration, and you will not be able to use it. You can find more information on this here - https://tailwindcss.com/docs/content-configuration#dynamic-class-names.

So, make sure to use complete class names. Alternatively, you can add the ones you're going to use as comments, although this is not a good practice due to maintainability issues.

**Reply by emiliaiwu** [reply]
@alex-krasikau Thanks for your answer! I appreciate it.

**Reply by MetodyPetrov** [reply]
@alex-krasikau 

> However, when I hardcode the color directly as bg-[#fffff], it works as expected.

Meaning tailwind is indeed being applied on the file, but the problem stays. And the issue persists even if in her example:

```
const color = userPreferences.color;
const hoverColorClass = `bg-[${color}]`;
const fullHover = `hover:${hoverColorClass}`;

return (
  <li
    key={index}
    className={`${fullHover} h-10 font-DMSans flex items-center justify-center relative group`}
  >
    {/* My list item content */}
  </li>
);
```
I am having the same issue after declaring an array of objects that I then array.map(callbackfn) and return a react Link component with a className prop depending on some calculations I do in the callbackfn. When giving that Link component a dynamic background class everything is fine but when I try to add a hover: class it's not working as expected.

**Comment by adamwathan** [comment]
Try this approach:

```jsx
const color = userPreferences.color;

return (
  <li
    key={index}
    style={{'--hover-color': color }}
    className={`hover:bg-[--hover-color] h-10 font-DMSans flex items-center justify-center relative group`}
  >
    {/* My list item content */}
  </li>
);
```

**Reply by emiliaiwu** [reply]
@adamwathan Yeah!! It worked! Thank you so much!!!

**Reply by shoaibmarif** [reply]
@adamwathan Thanks for the solution just adding a description for your solution.

In the following code we are using CSS variable to store the color first and then rendering it as a tailwind class.

**Reply by ejm** [reply]
As a heads up for anyone finding this in the future, [Tailwind CSS v4 changed the syntax for referencing the values of arbitrary variables](https://tailwindcss.com/docs/upgrade-guide#variables-in-arbitrary-values) from `[]` to `()`

The approach still works, just the brackets need to be swapped for parentheses:
```diff
- className={`hover:bg-[--hover-color] h-10 font-DMSans flex items-center justify-center relative group`}
+ className={`hover:bg-(--hover-color) h-10 font-DMSans flex items-center justify-center relative group`}
```

Hope this helps someone!
